### PR TITLE
wip: Docker-ize testing-workshop

### PR DESCRIPTION
**Pull request by edm00se**
This should address the inconsistencies between environments (e.g.- Windows 😕). Allows for running the app code in a Docker container, via docker-compose. Docker and docker-compose are far more normal of a dev dependency than instructing a person to roll an entire vm; at least IMO (some aren't phased by VMs, but more people seem to intuitively get 'install x, run y').

I've listed the highlights, steps to use, and some thoughts on process below. I'm looking for feedback regarding handling of hooking into the `validation` script, etc. For now, kicking off this PR!

### Requires
- Docker
- docker-compose
- Docker daemon(/app) running
- (Mac homebrew version `brew install docker docker-compose`)

### Steps to Use
- open CLI of preference (Terminal.app, LXTerm, cmd.exe, PowerShell...)
- cd to repo (no need to `npm install` or `npm run setup`)
- to launch with dev script running (`npm start dev`), run
```sh
docker-compose up
```

Note: to run other scripts, use `docker-compose run` with `--entrypoint "SCRIPT-TO-RUN"` and specify the name of the service, "workshop".

```sh
docker-compose run --entrypoint "npm start validate" workshop
```

### Shortcomings vs Benefits
The downside of throwing everything in a Docker container is that _everything_ is in the container. This is generally counteracted through my running `npm run setup` during the image build, so that a one-off run (`docker-compose run...`) or `docker-compose up` (defaults to run `npm start dev`) doesn't incur an re-run of setup or copying of source files (since the volume mounts from repo cwd to `/src` in the container).

#### MongoDB
I wanted initially to turn this into a "proper" service in the `docker-compose.yml`, but the issue with that was the fact that there are so many comparatively "hard coded" references to to checking for the `mongo` and `mongod` commands, etc. in validate, etc. ... go figure. Converting to this approach could happen, but would require some shimming around the setup, validation, and execution scripts that kick off the `mongod` process (and instead assume it's there, just connect to it).

#### Scripts
Ideally, there would be an easier way of running something other than `npm start dev`. We could include something into `package-scripts.js`, but that will assume `npm install` from the host... and the whole point of containerization here is to handle such fiddly things as dependencies internally to the container. Alternatively, this could be a part of the scripts in `package.json` and just alias the `docker-compose run...` command, but would avoid all the validation lib that's already there... if nps is installed. 🤔

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# node boron is 6(.x LTS) line, official image is the base
+FROM node:boron
+
+# needs xvfb to run correctly
```
Comment (edm00se) [inline Dockerfile]: That's about how that went. I'm glad people see that 😆 .

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+# make sure our PATH includes yarn and mongo(d)
+ENV PATH /mongodb-linux-x86_64-$MONGODB_VERSION/bin:/root/.yarn/bin:$PATH
+
+# we work out of the container's /src/ dir, copy in files to build
+WORKDIR /src
+COPY . /src
+# npm installs, then run setup script
+RUN npm install -g --quiet nodemon jest \
+  && npm install --quiet
```
Comment (edm00se) [inline Dockerfile]: While testing it locally, I had gotten an error re: jest, so I installed it globally and it magically "went away". The inclusion of nodemon was for some potentiality I didn't run into; likely not needed and should be removed.

Comment (kentcdodds) [inline Dockerfile]: Cool, sounds good. Let's not spend too much time worrying about that for now :)

**Comment by kentcdodds** [comment]
Yeah... Starting to think that the only real solution is to have people install via VirtualBox... Or maybe I'll just drop the cypress stuff and demo it instead :-(

**Comment by edm00se** [comment]
Cypress looks so fascinating and useful, it would be a shame to lose it entirely. Maybe as an opt-in feature?

**Comment by kentcdodds** [comment]
As a side-note, I'm incredibly frustrated at myself for not checking (and those who knew and didn't tell me) that Cypress does not work on Windows. That's SUCH a huge deal and I never would have started using Cypress for a workshop if I had known that. I can't really recommend that people use it until Cypress comes to them where they're at. Super frustrating.

**Comment by edm00se** [comment]
After that last round of issues I ran into, I assumed the Windows vm I was testing in was weirdly configured (it wouldn't be the first time) and kept moving. I wish I'd seen it for what it was as an issue :-|. I definitely think that once it rounds out on the supported OS, it'll be a game changer.

**Comment by edm00se** [comment]
As a side note, I would run `brew uninstall docker` on your machine and install again via [the Docker store's "desktop app"](https://store.docker.com/editions/community/docker-ce-desktop-mac?tab=description).

[The Windows version](https://store.docker.com/editions/community/docker-ce-desktop-windows?tab=description) of this gets trickier, as the equivalent (app-like) requires Windows 10 Pro or Enterprise (64-bit). For other/older versions, [docker toolbox](https://www.docker.com/products/docker-toolbox) is needed.

**Comment by edm00se** [comment]
Well, I'm at about a point where I can document accordingly and I _could_ have this ready as a proper feature PR, but we're back to "what's the gain?" Since the whole point was to encapsulate cypress in a known working OS and make it more friendly to the user, I think we may have _technically_ achieved this, without any of the benefit of the GUI from cypress, making it negligibly an enhancement, at best. Yes, it will "work", but we lose what we were hoping to gain. The documentation remaining is meant to be a capture of my reproducing on an Windows 10 machine; in limbo since my Win10 VM doesn't have enough memory for the moment.

Strictly speaking, it appears some crazy person(s) has been able to get an X11 app to run, exposing the containered graphical output through a combination of xvfb and running a "remote monitor" on the host (others have used vnc), but while I'm not certain, I expect that means a requirement of the host having an X11 compatible "monitor" environment. There have been other projects to make a GUI app/test suite run inside a Docker container, without exposing the X output, such as [chromium-xvfb](https://github.com/mark-adams/docker-chromium-xvfb) which allows for things like Karma testing in Chromium. That might be achievable, but probably not myself.

Considering the biggest shortcoming regarding cypress in Docker not having GUI access, I'm debating whether I should continue with this or not. I'm going to close it for now, unless things change or my understanding of what's possible evolves.

**Comment by kentcdodds** [comment]
Thank you so much for all your work on this. Hopefully we can get a vagrant thing set up so people have an easier experience setting up a virtual machine. Most people will be on linux-based environments (mac), so hopefully the impact is minimal. Thanks again, I really can't express my gratitude enough.

**Comment by edm00se** [comment]
You are quite welcome. You're always quite friendly and open about helping people in open source contributions, I was just attempting to return some of that good will. I'll keep an eye on cypress, it looks destined to be awesome.

**Pull request closed without merging**